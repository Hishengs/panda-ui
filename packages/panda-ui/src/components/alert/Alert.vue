<template>
  <!-- <panda-fade-transition> -->
  <div v-if="!closed" class="panda-alert" :class="cClass">
    <div v-if="showIcon && iconName" class="panda-alert-icon">
      <Icon :type="iconName" :size="20" />
    </div>
    <div class="panda-alert-main">
      <div v-if="title || $slots.title" class="panda-alert-title">
        <slot name="title">
          <h6>{{ title }}</h6>
        </slot>
      </div>
      <div class="panda-alert-content">
        <slot />
      </div>
    </div>
    <Close v-if="closable" class="panda-alert-close-btn" @click="onClose" />
  </div>
  <!-- </panda-fade-transition> -->
</template>

<script lang="ts" setup>
import Icon from '../icon';
import Close from './Close.vue';
import { computed, ref } from 'vue';

defineOptions({
  name: 'panda-alert',
});

const props = defineProps({
  type: {
    type: String,
    validator(val: string) {
      return ['success', 'info', 'warning', 'error'].includes(val);
    },
  },
  title: String,
  closable: Boolean,
  icon: String,
  showIcon: Boolean,
});

const closed = ref(false);
const statusIcon = computed(() => {
  const statusMap = {
    success: 'icon_feedback_success_fill',
    info: 'icon_tips_tips_fill',
    warning: 'icon_feedback_warning_fill',
    error: 'icon_feedback_failure_fill',
  };
  return statusMap[props.type as keyof typeof statusMap] || '';
});
const iconName = computed(() => props.icon || statusIcon);
const cClass = computed(() => {
  return {
    [`panda-alert-closable`]: props.closable,
    [`panda-alert-${props.type}`]: !!props.type,
  };
});

const emit = defineEmits({
  close: (e: MouseEvent) => e instanceof MouseEvent,
});

const onClose = (e: MouseEvent) => {
  closed.value = true;
  emit('close', e);
};
</script>
