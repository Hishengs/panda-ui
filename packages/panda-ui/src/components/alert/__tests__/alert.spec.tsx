import { describe, it, expect, beforeEach } from 'vitest';
import { mount, shallowMount, VueWrapper } from '@vue/test-utils';
import Alert from '../index';

describe('Alert Component', () => {
  /* it('should be regiesterd as a plugin', () => {
    const localVue = createLocalVue();
    localVue.use(Alert);
    const { components } = localVue.options;
    expect(typeof components[Alert.name]).toBe('function');
  }); */

  describe('Default Props', () => {
    let wrapper: VueWrapper;
    beforeEach(() => {
      wrapper = shallowMount(Alert);
    });

    it('should type be one of ["success", "info", "warning", "error"], default is undefined', () => {
      const type = wrapper.vm.$options.props.type;

      expect(type.type).toBe(String);
      expect(type.default).toEqual(undefined);
      expect(type.validator && type.validator('success')).toBeTruthy();
      expect(type.validator && type.validator('info')).toBeTruthy();
      expect(type.validator && type.validator('warning')).toBeTruthy();
      expect(type.validator && type.validator('error')).toBeTruthy();
      expect(type.validator && type.validator('foo')).toBeFalsy();
    });

    // it('should align be one of ["left", "center", "right"]', () => {
    //   const align = wrapper.vm.$options.props.align;

    //   expect(align.type).toBe(String);
    //   expect(align.validator && align.validator('left')).toBeTruthy();
    //   expect(align.validator && align.validator('center')).toBeTruthy();
    //   expect(align.validator && align.validator('right')).toBeTruthy();
    //   expect(align.validator && align.validator('foo')).toBeFalsy();
    // });

    it('should not render alert-Icon when showIcon is false by default', () => {
      expect(wrapper.find('.panda-alert-icon').exists()).toBeFalsy();
    });

    it('should not render title when title is empty string by default', () => {
      expect(wrapper.find('.panda-alert-title').exists()).toBeFalsy();
    });

    it('should not render close-Icon when closable is false by default', () => {
      expect(wrapper.find('.panda-alert-close-btn').exists()).toBeFalsy();
    });
  });

  describe('Custom Props', () => {
    let wrapper: VueWrapper;
    beforeEach(() => {
      wrapper = mount(Alert);
    });

    it('should render icon when showIcon is true', async () => {
      await wrapper.setProps({
        type: 'success',
        showIcon: true,
      });

      expect(wrapper.find('.panda-alert-icon').exists()).toBeTruthy();
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should not render icon when showIcon is false', async () => {
      await wrapper.setProps({
        showIcon: false,
      });

      expect(wrapper.find('.panda-alert-icon').exists()).toBeFalsy();
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render title <h6>foo</h6> when title is foo', async () => {
      await wrapper.setProps({
        title: 'foo',
      });

      expect(wrapper.find('.panda-alert-title h6').html()).toBe('<h6>foo</h6>');
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render close-Icon when closable is true', async () => {
      await wrapper.setProps({
        closable: true,
      });

      expect(wrapper.find('.panda-alert-close-btn').exists()).toBeTruthy();
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('Slot', () => {
    it('should render <p>foo</p> when default slot is <p>foo</p>', async () => {
      const wrapper = await mount(Alert, {
        slots: {
          default: '<p>foo</p>',
        },
      });

      expect(wrapper.find('.panda-alert-content p').html()).toBe('<p>foo</p>');
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render title <h2>foo</h2> when title-slot is <h2>foo</h2>', async () => {
      const wrapper = await mount(Alert, {
        slots: {
          title: '<h2>foo</h2>',
        },
      });

      expect(wrapper.find('.panda-alert-title h2').html()).toBe('<h2>foo</h2>');
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('Event', () => {
    it('should remove alert when close-Icon is clicked', async () => {
      const wrapper = mount(Alert, {
        props: {
          closable: true,
        },
      });

      expect(wrapper.find('.panda-alert').exists()).toBe(true);
      await wrapper.find('.panda-alert-close-btn').trigger('click');
      expect(wrapper.find('.panda-alert').exists()).toBe(false);
    });
  });
});
