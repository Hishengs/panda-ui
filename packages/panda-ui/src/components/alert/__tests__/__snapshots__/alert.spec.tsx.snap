// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Alert Component > Custom Props > should not render icon when showIcon is false 1`] = `
"<!-- <panda-fade-transition> -->
<div class=\\"panda-alert\\">
  <!--v-if-->
  <div class=\\"panda-alert-main\\">
    <!--v-if-->
    <div class=\\"panda-alert-content\\"></div>
  </div>
  <!--v-if-->
</div>
<!-- </panda-fade-transition> -->"
`;

exports[`Alert Component > Custom Props > should render close-Icon when closable is true 1`] = `
"<!-- <panda-fade-transition> -->
<div class=\\"panda-alert panda-alert-closable\\">
  <!--v-if-->
  <div class=\\"panda-alert-main\\">
    <!--v-if-->
    <div class=\\"panda-alert-content\\"></div>
  </div><svg width=\\"16\\" height=\\"16\\" viewBox=\\"0 0 16 16\\" fill=\\"currentColor\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"panda-alert-close-btn\\">
    <path fill-rule=\\"evenodd\\" clip-rule=\\"evenodd\\" d=\\"M13.7803 2.21967C13.4874 1.92678 13.0126 1.92678 12.7197 2.21967L8 6.93934L3.28033 2.21967L3.19621 2.14705C2.9026 1.9292 2.48594 1.9534 2.21967 2.21967C1.92678 2.51256 1.92678 2.98744 2.21967 3.28033L6.93934 8L2.21967 12.7197L2.14705 12.8038C1.9292 13.0974 1.9534 13.5141 2.21967 13.7803C2.51256 14.0732 2.98744 14.0732 3.28033 13.7803L8 9.06066L12.7197 13.7803L12.8038 13.8529C13.0974 14.0708 13.5141 14.0466 13.7803 13.7803C14.0732 13.4874 14.0732 13.0126 13.7803 12.7197L9.06066 8L13.7803 3.28033L13.8529 3.19621C14.0708 2.9026 14.0466 2.48594 13.7803 2.21967Z\\" fill=\\"currentColor\\"></path>
  </svg>
</div>
<!-- </panda-fade-transition> -->"
`;

exports[`Alert Component > Custom Props > should render icon when showIcon is true 1`] = `
"<!-- <panda-fade-transition> -->
<div class=\\"panda-alert panda-alert-success\\">
  <div class=\\"panda-alert-icon\\">
    <!--v-if-->
  </div>
  <div class=\\"panda-alert-main\\">
    <!--v-if-->
    <div class=\\"panda-alert-content\\"></div>
  </div>
  <!--v-if-->
</div>
<!-- </panda-fade-transition> -->"
`;

exports[`Alert Component > Custom Props > should render title <h6>foo</h6> when title is foo 1`] = `
"<!-- <panda-fade-transition> -->
<div class=\\"panda-alert\\">
  <!--v-if-->
  <div class=\\"panda-alert-main\\">
    <div class=\\"panda-alert-title\\">
      <h6>foo</h6>
    </div>
    <div class=\\"panda-alert-content\\"></div>
  </div>
  <!--v-if-->
</div>
<!-- </panda-fade-transition> -->"
`;

exports[`Alert Component > Slot > should render <p>foo</p> when default slot is <p>foo</p> 1`] = `
"<!-- <panda-fade-transition> -->
<div class=\\"panda-alert\\">
  <!--v-if-->
  <div class=\\"panda-alert-main\\">
    <!--v-if-->
    <div class=\\"panda-alert-content\\">
      <p>foo</p>
    </div>
  </div>
  <!--v-if-->
</div>
<!-- </panda-fade-transition> -->"
`;

exports[`Alert Component > Slot > should render title <h2>foo</h2> when title-slot is <h2>foo</h2> 1`] = `
"<!-- <panda-fade-transition> -->
<div class=\\"panda-alert\\">
  <!--v-if-->
  <div class=\\"panda-alert-main\\">
    <div class=\\"panda-alert-title\\">
      <h2>foo</h2>
    </div>
    <div class=\\"panda-alert-content\\"></div>
  </div>
  <!--v-if-->
</div>
<!-- </panda-fade-transition> -->"
`;
